### PR TITLE
Regulator data

### LINKED ISSUE #1283: Create regulator data stream
Provide mechanism to deliver daily dump of filing data (`Transmittal Sheet` and `Loan Application Register`) to regulators. High level requirements are as follows:

- Files will be automatically uploaded to an `Amazon Web Services S3` bucket
- `Loan Application Register` data is to be extracted daily at 10PM and made available as a single pipe delimited `csv` file, uncompressed, following the same format used at ingestion time as per the [Filing Instruction Guide](https://www.consumerfinance.gov/data-research/hmda/static/for-filers/2017/2017-HMDA-FIG.pdf). File naming convention will include time and date stamp following the [`ISO-8601`](https://en.wikipedia.org/wiki/ISO_8601) international standard.
- `Transmittal Sheet` data is to be extracted daily at 10PM and made available as a single pipe delimited `csv` file, uncompressed, following the same format used at ingestion time as per the [Filing Instruction Guide](https://www.consumerfinance.gov/data-research/hmda/static/for-filers/2017/2017-HMDA-FIG.pdf). In addition to this, a field will be appended at the end of each line (representing a single Financial Institution), with the date of latest filing. File naming convention will include time and date stamp following the [`ISO-8601`](https://en.wikipedia.org/wiki/ISO_8601) international standard.
- `Panel` data is to be extracted daily at 10PM and made available as a single pipe delimited `csv` file, uncompressed, following the same format used at ingestion time. 
- These processes should also be available on demand.

### LINKED ISSUE #1342: Produce Panel file and schema for FRB transfer
The FRB has requested an update of the Panel file whenever changes are made to to the data.

The Panel file will need a transfer schema, a Platform produced file, and will be zipped and named CFPB_HMDA_panel.zip.

To date, the Panel has been transferred as a pipe-delimited text file that used the HMDA Panel schema for Panel data given to Platform.

The FRB requires all LAR and TS records (identified by the concatenation of Agency Code and Respondent ID) are present in the Panel file. To fulfill this requirement, all Agency Code and Respondent IDs received from filers need to either match the fields or be made consistent, potentially by overwriting the Panel field. Panel issue is [here](https://github.com/cfpb/hmda-platform/issues/1342).

The requested delivery mechanism is the same as for the TS and LAR files.

### PR DESCRIPTION
**Pull request by jmarin**
Closes #1283

### CODE REVIEW

File: publication/src/main/scala/hmda/publication/regulator/lar/ModifiedLarPublisher.scala [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+import hmda.query.repository.filing.LoanApplicationRegisterCassandraRepository
+
+import scala.concurrent.duration._
+
+object ModifiedLarPublisher {
+  val name = "SubmissionSignedModifiedLarSubscriber"
+  def props(supervisor: ActorRef): Props = Props(new ModifiedLarPublisher(supervisor))
+}
+
+class ModifiedLarPublisher(supervisor: ActorRef) extends HmdaActor with LoanApplicationRegisterCassandraRepository {
```
Comment (schbetsy) [inline publication/src/main/scala/hmda/publication/regulator/lar/ModifiedLarPublisher.scala]: It looks like `ModifiedLarPublisher`, `RegulatorLarPublisher`, `RegulatorPanelPublisher`, and `RegulatorTsPublisher` duplicate a bunch of configuration code. Could they share that code in a common module instead?

File: query/src/main/scala/hmda/query/projections/filing/TransmittalSheetSignedEventSubscriber.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  val sink = CassandraSink[TransmittalSheetWithTimestamp](parallelism = 2, preparedStatement, statementBinder)
+
+  override def receive: Receive = {
+
+    case SubscribeAck(Subscribe(PubSubTopics.submissionSigned, None, `self`)) =>
+      log.info(s"${self.path} subscribed to ${PubSubTopics.submissionSigned}")
+
+    case SubmissionSignedPubSub(submissionId) =>
+      log.info(s"${self.path} received submission signed event with submission id: ${submissionId.toString}")
+      val perssitenceId = s"HmdaFileValidator-$submissionId"
```
Comment (schbetsy) [inline query/src/main/scala/hmda/query/projections/filing/TransmittalSheetSignedEventSubscriber.scala]: Typo in "persistence".

File: publication/src/main/scala/hmda/publication/regulator/lar/RegulatorLarPublisher.scala [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            "purchaser_type|" +
+            "denial1|" +
+            "denial2|" +
+            "denial3|" +
+            "rate_spread|" +
+            "hoepa_status|" +
+            "lien_status\n"
+        ).toIterator).map(s => ByteString(s))
+
+      val larSource = readData(fetchSize)
```
Comment (schbetsy) [inline publication/src/main/scala/hmda/publication/regulator/lar/RegulatorLarPublisher.scala]: Question: Are we uploading **every** LAR to the regulator bucket every day? Like, if a LAR file gets uploaded on Jan 1, will it be included in the dump every day of the filing season?

If there's logic to limit the dump to certain LARS (e.g. the ones uploaded on a certain day), I'm missing it, and I'd appreciate it if you point it out.

Comment (jmarin) [reply publication/src/main/scala/hmda/publication/regulator/lar/RegulatorLarPublisher.scala]: We are loading every LAR, this is by design (what they asked for)

File: publication/src/main/scala/hmda/publication/regulator/ts/RegulatorTsPublisher.scala [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import com.typesafe.akka.extension.quartz.QuartzSchedulerExtension
+import hmda.persistence.model.HmdaActor
+import hmda.publication.regulator.messages.PublishRegulatorData
+import hmda.query.repository.filing.TransmittalSheetCassandraRepository
+
+object RegulatorTsPublisher {
+  def props(): Props = Props(new RegulatorTsPublisher)
+}
+
+class RegulatorTsPublisher extends HmdaActor with TransmittalSheetCassandraRepository {
```
Comment (schbetsy) [inline publication/src/main/scala/hmda/publication/regulator/ts/RegulatorTsPublisher.scala]: I made this comment previously, but it looks like it got eaten: It looks like `ModifiedLarPublisher`, `RegulatorLarPublisher`, `RegulatorPanelPublisher`, and `RegulatorTsPublisher` duplicate a bunch of configuration code. Could they share that code in a common module instead?

Comment (jmarin) [reply publication/src/main/scala/hmda/publication/regulator/ts/RegulatorTsPublisher.scala]: Yeah, forgot to reply to that comment. They could, but I encountered some issues while doing that refactor and decided against it at this time

**Pull request merged**